ludwig/schema/hyperopt/scheduler.py

```diff
@@ -0,0 +1,509 @@
+            " each trial as it finishes one step of training. The callable must take four arguments: TrialRunner, "
+            "current Trial, current result dict and the ResourceChangingScheduler calling it. The callable must return "
+            "a PlacementGroupFactory, Resources, dict or None (signifying no need for an update). If "
+            "resources_allocation_function is None, no resource requirements will be changed at any time. By default, "
+            "DistributeResources will be used, distributing available CPUs and GPUs over all running trials in a robust"
+            " way, without any prioritization."
+        )
+    )
+
+
+def get_scheduler_conds():
+    """Returns a JSON schema of conditionals to validate against scheduler types defined in
+    `ludwig.schema.hyperopt.scheduler_registry`."""
+    conds = []
+    for scheduler_config in scheduler_config_registry:
+        scheduler_cls = scheduler_config_registry[scheduler_config]
+        other_props = schema_utils.unload_jsonschema_from_marshmallow_class(scheduler_cls)["properties"]
+        schema_utils.remove_duplicate_fields(other_props)
+        preproc_cond = schema_utils.create_cond(
+            {"type": scheduler_config},
+            other_props,
+        )
+        conds.append(preproc_cond)
+    return conds
+
+
+def SchedulerDataclassField(default={"type": "fifo"}, description="Hyperopt scheduler settings."):
+    """Custom dataclass field that when used inside of a dataclass will allow any scheduler in
+    `ludwig.schema.hyperopt.scheduler.scheduler_registry`. Sets default scheduler to 'fifo'.
+
+    :param default: Dict specifying a scheduler with a `type` field and its associated parameters. Will attempt to use
+           `type` to load scheduler from registry with given params. (default: {"type": "fifo"}).
+    :return: Initialized dataclass field that converts untyped dicts with params to scheduler dataclass instances.
+    """
+
+    class SchedulerMarshmallowField(fields.Field):
+        """Custom marshmallow field that deserializes a dict to a valid scheduler from
+        `ludwig.schema.hyperopt.scheduler_registry` and creates a corresponding `oneOf` JSON schema for external
+        usage."""
+
+        def _deserialize(self, value, attr, data, **kwargs):
+            if value is None:
+                return None
+            if isinstance(value, dict):
+                if "type" in value and value["type"] in scheduler_config_registry:
+                    scheduler_config_cls = scheduler_config_registry[value["type"].lower()]
+                    try:
+                        return scheduler_config_cls.Schema().load(value)
+                    except (TypeError, ValidationError) as e:
+                        raise ValidationError(
+                            f"Invalid params for scheduler: {value}, see `{opt}` definition. Error: {e}"
+                        )
+                raise ValidationError(
```

Comment:
@arnavgarg1 As you noticed in a prior comment (https://github.com/ludwig-ai/ludwig/pull/2346#discussion_r1028048497), this schema as configured would prevent empty scheduler configs. Like I said, I feel that we should restrict this as such but can I get double confirmation? Otherwise I can change it so that empty configs are accepted.

Reply:
I think it might be ok to prevent empty scheduler configs, but I think there are two main things to keep in mind:
1. We should update the Ludwig docs to ensure that users know it is required, and also list out all of the options on Ludwig docs itself instead of users having to go to Ray docs.
2. (will message you)

Reply:
Is my understanding correct that users must specify the scheduler config, but we also have it default to Fifo? If it must be specified, does it make sense to have a default at all?

Reply:
@arnavgarg1 It's a bit more subtle. Specifically a user cannot set `scheduler: None` - ergo if they set a `scheduler` config they have to go all the way with the configuration. If they choose not to specify it at all then the default `scheduler` is chosen. This is the same thing we do for the optimizer, for example.

Reply:
@arnavgarg1 yup